File: LeetCode/542. 01 Matrix_bfs.java
class Solution {
    int[] DIR = new int[]{0, 1, 0, -1, 0};
    public int[][] updateMatrix(int[][] m) {
        int rows = m.length;
        int cols = m[0].length;

        LinkedList<int[]> q = new LinkedList<>();
        for (int r = 0; r < rows; r++) {
            for (int c = 0; c < cols; c++) {
                if (m[r][c] == 0) q.push(new int[] {r, c});
                else m[r][c] = -1;
            }
        }
        while (!q.isEmpty()) {
            int[] cur = q.poll();
            int r = cur[0];
            int c = cur[1];
            for (int i = 0; i<4; i++) {
                int nr = r + DIR[i], nc = c + DIR[i + 1];
                if (nr < 0 || nr == rows ||
                    nc < 0 || nc == cols ||
                    m[nr][nc] != -1) continue;
                m[nr][nc] = m[r][c] + 1;
                q.add(new int[]{nr, nc});
            }
        }
        return m;
    }
} //TC: O(m*n), SC: O(m*n)

File: LeetCode/542. 01 Matrix_bfs.kt
class Solution {
    fun updateMatrix(m: Array<IntArray>): Array<IntArray> {
        val q = LinkedList<IntArray>()

        for (r in 0 until m.size) {
            for (c in 0 until m[0].size) {
                if (m[r][c] == 0) { // check if it is sea, queue if it is
                    q.add(intArrayOf(r, c))
                } else m[r][c] = -1 // if not, mark it as visited by assigning -1
            }
        }

        while (q.isNotEmpty()) { // loop until the queue is empty
            val c = q.poll()
            val x = c[0]
            val y = c[1]

            for (d in arrayOf(
                arrayOf(1, 0),
                arrayOf(-1, 0),
                arrayOf(0, 1),
                arrayOf(0, -1))
                ) {
                val tmpX = x + d[0]
                val tmpY = y + d[1] // loop each direction

                if (tmpX >= 0 && tmpX < m.size &&
                    tmpY >= 0 && tmpY < m[0].size &&
                    m[tmpX][tmpY] == -1) { // if the coordinate is within bounds and is a land (-1)

                    m[tmpX][tmpY] = m[x][y] + 1 // assign the land to be current coordinate value + 1
                    q.add(intArrayOf(tmpX, tmpY)) // add the land to queue
                }
            }
        }
        return m
    }
}
